app/models/semester_day.rb

```diff
@@ -0,0 +1,64 @@
+require 'day'
+require 'faculty_semester'
+
+class SemesterDay
+
+  attr_reader :date, :semester
+
+  def initialize(semester, date)
+    @semester = semester.freeze
+    @date = date.freeze
+  end
+
+  def self.resolve(date, faculty_id)
+    if (semester = FacultySemester.find_by_date(date, faculty_id))
+      new semester, date
+    end
+  end
+
+  def wday
+    return date.wday unless period.first_day_override
+
+    @wday ||= begin
+      offset = Day.to_numeric(period.first_day_override) - period.starts_at.wday
+      (date.wday + offset) % 7
+    end
+  end
+
+  def wday_name
+    Day.from_numeric(wday)
+  end
+
+  def week_parity
+    return if period.type != :teaching
+
+    @week_parity ||= begin
+      first_parity = period.first_week_parity == :even ? 0 : 1
+      parity_name(period.starts_at.cweek + date.cweek + first_parity)
+    end
+  end
+
+  def period
+    @period ||= @semester.semester_periods.find do |p|
+      date >= p.starts_at  && date <= p.ends_at
+    end
+  end
+
+  def week_num
+    @week_num ||= begin
+      periods = @semester.semester_periods.find_all do |p|
+        p.type == period.type
+      end
+      cweeks = periods.reduce([]) do |acc, p|
+        acc | (p.starts_at.cweek..p.ends_at.cweek).to_a
+      end
+      cweeks.index(date.cweek) + 1
```

Comment:
:white_check_mark: 35d11222451043fda31e28dcaba0bcbb5f46c334 (efficiency)

I’m sorry that it’s hard to read for you, but this is the most clean and elegant implementation from all I’ve been thinking about. If you have a better one, please propose it.

Look at 6050de39c4738513125c67a1438f84b306a7a056, I’ve added some comments and tests in 1ac5a8441acc1a910f970782a32c21f07b2abfce.